### PR TITLE
[Bugfix] Trigger failed build

### PR DESCRIPTION
**Pull request by schultek**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).

### Motivation and Context

The 'Trigger Failed Build' Button did not work properly as described in #2329 

### Description

- Fix the BambooBuildPlanDTO to use the correct property names.
- Add additional checks the the trigger-failed-build handler to support edge-cases.

### Steps for Testing

--- 
**Repeat the following steps:**
- For exercises with & without due-dates and for submissions before and after the due-date
- From inside the code-editor, from the exercise details page and as an instructor from the participations page.
---

1. Log in to Artemis
2. Create a Programming Exercise
3. Modify the Build Plan to take at least 3 minutes (use `sleep 180` in the script)
    - Peek 2020-11-04 15-32
4. Participate in the exercise, submit and wait for the building indicator to timeout.
5. Click the Trigger Build Button while the Build is still running.
    - Bildschirmfoto 2020-11-07 um 01 37 39
6. You should again see the building indicator without any errors.
    - Bildschirmfoto 2020-11-07 um 01 28 31

### Test Coverage

- Nothing changed

### Screenshots

No UI changes

### CODE REVIEW

**Review by tiffca — APPROVED** [review]
I don't see the button to trigger the build, but submitting the exercise again triggers the build fine. The build works fine and no issues encountered while participating in exercise. 
for example this exercise on test server 1: 
https://artemistest.ase.in.tum.de/#/courses/304/programming-exercises/4536/code-editor/231747


image

**Comment by stefanwaldhauser** [comment]
Testing it locally
Run Configuration: 
image
image
image
I can confirm that i do not get the build timeout? I tested it locally.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java [hunk 1]
```diff
@@ -186,7 +187,9 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
             programmingSubmissionService.notifyUserAboutSubmission(submission.get());
             return ResponseEntity.ok().build();
         }
-        if (lastGraded) {
+        if (lastGraded && submission.get().getType() != SubmissionType.INSTRUCTOR && submission.get().getType() != SubmissionType.TEST
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:190]: Could you add a explanation here? This is not that intuitive to understand

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java [hunk 2]
```diff
@@ -186,7 +187,9 @@ public ProgrammingSubmissionResource(ProgrammingSubmissionService programmingSub
...
+                && submission.get().getParticipation().getExercise().getDueDate() != null
+                && submission.get().getParticipation().getExercise().getDueDate().isBefore(ZonedDateTime.now())) {
```
Comment (fde312) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:192]: Could you add a explanation here? This is not that intuitive to understand

**Review by MaisiKoleni — DISMISSED** [review]
I have not tried all variants, but the main problem seems to be fixed. Code looks good to me, but I would appreciate a comment why the condition using lastGraded looks like that.

**Comment by fde312** [comment]
@kloessst and I encountered some inconsistencies:
the check that you extended with the request parameter`lastGraded` is called in the code editor only with value `false`. This is intended right?

**Review by kloessst — APPROVED** [review]
Tested without due date, with due date and over the instructor participations view.
The changes are reasonable. 
- Before `notFound()` was always returned if no build was running currently and `lastGraded` was true. `lastGraded` is set to true for trigger failed build buttons displayed in views which only show graded results.
- Now `notFound()` is only returned if `lastGraded` is true, no build currently runs on the CI, a student submitted and the due date passed. I could reproduce this case by setting a due date 1 minute into the future, submitting, canceling the build, waiting for the re-trigger build button to appear and finally submitting the button of the due date passed.

Took a while for @fde312 and me to figure this out 😃

**Review by fde312 — DISMISSED** [review]
The main problem was fixed that the not found error was thrown always when the retrigger button was called e.g. from the exercise details view. Good job! 👍

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java (hunk 2 above)
Comment (schultek) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingSubmissionResource.java:192]: Done

**Review by kloessst — APPROVED** [review]
Re-approve

**Pull request merged**